### PR TITLE
Don't overwrite local changes when updating submodules

### LINKED ISSUE #103485: bootstrap submodule sync overwrites local changes
When bootstrap updates a submodule, it performs a hard reset, discarding all local changes: https://github.com/rust-lang/rust/blob/1481fd964bac3c750c7e1b21206fdaa60346c456/src/bootstrap/lib.rs#L631-L632

This was a nasty surprise when I was editing my local `src/llvm-project`, and then did a `git pull` and `./x.py build`.
Fortunately, I only had some extra debugging prints added, but there was no indication that my local changes would be overwritten.

We should require explicit confirmation before potentially deleting local changes like this.

### PR DESCRIPTION
**Pull request by dotdot0**
Fixes https://github.com/rust-lang/rust/issues/103485

### CODE REVIEW

**Comment by jyn514** [comment]
r? @jyn514 

@pratushrai0309 can you please fix the tidy error?

File: src/bootstrap/lib.rs [hunk 1]
```diff
@@ -647,9 +647,11 @@ impl Build {
         if !update(true).status().map_or(false, |status| status.success()) {
             self.run(&mut update(false));
         }
-
+        // 
+        self.run(Command::new("git").args(&["stash", "push"]).current_dir(&absolute_path));
         self.run(Command::new("git").args(&["reset", "-q", "--hard"]).current_dir(&absolute_path));
-        self.run(Command::new("git").args(&["clean", "-qdfx"]).current_dir(absolute_path));
+        self.run(Command::new("git").args(&["clean", "-qdfx"]).current_dir(&absolute_path));
+        self.run(Command::new("git").args(&["stash", "pop"]).current_dir(absolute_path))
```
Comment (jyn514) [inline src/bootstrap/lib.rs:654]:
```suggestion
        self.run(Command::new("git").args(&["stash", "pop"]).current_dir(absolute_path));
```

**Comment by pratushrai0309** [comment]
@jyn514  I think it's done

**Pull request merged**